sqlmesh/core/state_sync/engine_adapter.py

```diff
@@ -402,6 +425,109 @@ def _get_environment(
         env = self._environment_from_row(row)
         return env
 
+    @transactional()
+    def add_interval(
+        self,
+        snapshot: Snapshot,
+        start: TimeLike,
+        end: TimeLike,
+        is_dev: bool = False,
+    ) -> None:
+        logger.info("Adding interval for snapshot %s", snapshot.snapshot_id)
+
+        self.engine_adapter.insert_append(
+            self.intervals_table,
+            _intervals_to_df([snapshot], start, end, is_dev, False),
+            columns_to_types=self._interval_columns_to_types,
+        )
+
+    @transactional()
+    def remove_interval(
+        self,
+        snapshots: t.Iterable[SnapshotInfoLike],
+        start: TimeLike,
+        end: TimeLike,
+        all_snapshots: t.Optional[t.Iterable[Snapshot]] = None,
+    ) -> None:
+        all_snapshots = all_snapshots or self._get_snapshots_with_same_version(snapshots)
+
+        if logger.isEnabledFor(logging.INFO):
+            snapshot_ids = ", ".join([str(s.snapshot_id) for s in snapshots])
+            logger.info("Removing interval for snapshots: %s", snapshot_ids)
+
+        self.engine_adapter.insert_append(
+            self.intervals_table,
+            _intervals_to_df(all_snapshots, start, end, False, True),
+            columns_to_types=self._interval_columns_to_types,
+        )
+
+    @transactional()
+    def compact_intervals(self) -> None:
+        query = (
+            exp.select(
+                "id", "name", "identifier", "version", "start_ts", "end_ts", "is_dev", "is_removed"
+            )
+            .from_(self.intervals_table)
+            .join(
+                exp.select("name", "identifier")
+                .from_(self.intervals_table)
+                .where(exp.column("is_compacted").not_())
+                .distinct(),
+                using=["name", "identifier"],
+            )
+            .order_by(
+                "name", "identifier", "is_dev", "created_ts", "is_removed", "start_ts", "end_ts"
+            )
+        )
+
+        rows = self.engine_adapter.fetchall(query)
+        interval_ids = {row[0] for row in rows}
+        snapshot_intervals = _rows_to_snapshot_intervals(rows)
+
+        self._push_intervals(snapshot_intervals)
+
+        if interval_ids:
+            self.engine_adapter.delete_from(
+                self.intervals_table, exp.column("id").isin(*interval_ids)
+            )
+
+    def get_snapshot_intervals(
```

Comment:
maybe this takes in where filter and we can dry up previous method

Reply:
Refactored the code to reuse the query.